Improved mattermost export and import process for missing files

Nityanand13 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
 During bulk export process we have added a second log file which will contain the information of missing log file, since we do not want to force the user to download and unzip the export.zip file as it may be very large. 
During import process  if there are any missing files, then we have modified the logic to post a warning message for missing files in the server log but allowed the import to continue.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note

```

mattermod commented:
Hello @Nityanand13,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

isacikgoz commented:
Hey @Nityanand13 thanks for the contribution! I'll have a more focused look this after you update the tests. And also I'd like to hear the impact if this is going to have any impact to the export -> import process. Can you elaborate more on that? Would you like to add an e2e test to verify there is no impact on the export -> import workflow?

Apart from these, I have another question; would it make more sense to download the log file separately only when it's requested? Just something to discuss.

noxer commented:
I'm taking a look as well, sorry for the delay.

Nityanand13 commented:
> Hey @Nityanand13 thanks for the contribution! I'll have a more focused look this after you update the tests. And also I'd like to hear the impact if this is going to have any impact to the export -> import process. Can you elaborate more on that? Would you like to add an e2e test to verify there is no impact on the export -> import workflow?
> 
> Apart from these, I have another question; would it make more sense to download the log file separately only when it's requested? Just something to discuss.

Hii @isacikgoz  
Earlier during the export process, If any attachments were not available on disk it breaks the export process with an error. Now we have modified a logic as follows
1. Post a warning message for missing files in the server log but allow the export to continue.
2. Export a separate log file that lists the missing files.

Earlier during the import process, if there were any missing files it breaks the import process with an error. Now we have modified a logic as follows
1. Post a warning message for missing files in the server log but allow the import to continue.

Regarding your second question,
Actually, we are downloading the log file with a different command. For eg:
1. To download a zip file we can run the command: `$ mmctl export download 1xmmqdazapym8pn1ip7k6rn83e_export.zip`
2. To download the log file we can run the command: `$ mmctl export download 1xmmqdazapym8pn1ip7k6rn83e_export.log`

mkdbns commented:
@isacikgoz , following up on @Nityanand13 comment. Given that the download command requires a file name to download, it doesn't quite make sense to download the two files together. This change attempts to keep it simple by allowing the user to download what he needs. We were asked to not include the log file into the zip, keeping them separate as there may be different users/needs for the log file vs the full export zip.

sadohert commented:
@nab-77 - What are your thoughts here?  Are you good with the approach where separate files are available for download?

Makes sense to me.

isacikgoz commented on app/export.go:
```diff
@@ -63,82 +64,143 @@ var exportablePreferences = map[imports.ComparablePreference]string{{
+	if logWriter != nil && len(missingAttachments) != 0 {
+		fileData, err := json.MarshalIndent(imports.MissingAttachments{MissingAttachments: missingAttachments}, "", " ")
+		if err != nil {
+			return model.NewAppError("exportLogFile", "Failed to marshal and indent the data", nil, err.Error(), http.StatusInternalServerError)
```
We generally use error id and translates here. You can check other occurrences in the code.

isacikgoz commented on app/export.go:
```diff
@@ -554,7 +632,14 @@ func (a *App) exportCustomEmoji(writer io.Writer, outPath, exportDir string, exp
 				filePath = filepath.Join("emoji", emoji.Id, "image")
-				emojiPaths = append(emojiPaths, filePath)
+				emojiDataList = append(emojiDataList, &imports.AttachmentImportData{
+					Id:        emoji.Id,
+					CreatorId: emoji.CreatorId,
+					CreateAt:  emoji.CreateAt,
+					UpdateAt:  emoji.UpdateAt,
+					Path:      &filePath,
+					Name:      emoji.Name,
```
What's the reason we are adding fields like `Id`? They are not taken into account while importing right?

mkdbns replied:
@isacikgoz , we want to include the full record of the attachment in the export log so the user has everything they need to address or investigate the issue. This was requested/agreed upon by @nab-77 as we were working out the approach.

isacikgoz commented on jobs/export_process/worker.go:
```diff
@@ -36,28 +36,43 @@ func MakeWorker(jobServer *jobs.JobServer, app AppIface) model.Worker {
-		errCh := make(chan *model.AppError, 1)
 		go func() {
-			defer close(errCh)
-			_, appErr := app.WriteFile(rd, filepath.Join(outPath, exportFilename))
+			_, appErr := app.WriteFile(zipReader, filepath.Join(outPath, exportZipFilename))
```
How do we wait this goroutine? Since the defer function in `L:46` called right after `MakeWorker` function finished and if we can't finish `app.WriteFile`, we will try to send the error to a closed channel.

Nityanand13 replied:
But the `MakeWorker` function will not be finished until the `app.WriteFile` is finished. This is because when you look at `L:72`,  where we are waiting until we receive 2 values from the channel.

isacikgoz reviewed (changes requested):
@Nityanand13 Thanks for the update, logic seems okay to me. I'll review once again after the tests are fixed. Let me know if you need any help regarding to tests.

agnivade commented:
Thank you for the context @mkdbns ! I will sync up with Neil and Stu on this.

Nityanand13 commented on app/export.go:
```diff
@@ -63,82 +64,143 @@ var exportablePreferences = map[imports.ComparablePreference]string{{
+			if logData, err = a.ExportAttachments(attachment, outPath, zipWr); err != nil {
 				return err
 			}
+			if logData != nil {
+				missingAttachments = append(missingAttachments, logData)
+			}
```
I think it does not look good to append `nil` in the log file for the attachments that are available on disk. We can simply ignore that. What do you say?

agnivade commented:
@mkdbns / @Nityanand13 - I synced up with @nab-77 and here are the takeaways:
1. For export, the admin should run an integrity check first, and take steps to fix their installation. That might mean remounting volumes or cleaning up old posts with broken attachments.
2. For import, the mmctl import validator will auto-fix posts with broken attachments.
This should fix both the export and import paths.

Apologies to have wasted both of your time with this. In future, we will ensure to sync up internally before asking any feature like this.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!
